### ISSUE TITLE
Replace JSON.NET with System.Text.Json

### ISSUE THREAD

**Issue by rdelgado-incomm** [rebus-org/Rebus#1010]
Hello, we've done some work in our organization to replace Newtonsoft.Json with System.Text.Json.  We've found that Microsoft's implementation of serialization is more performant than Newtonsoft.Json.  

Would you be willing to entertain a pull-request that utilizes System.Text.Json instead?

**Comment by rdelgado-incomm** [comment]
I just went through the repo and noticed both are actually being referenced. Why is that?

**Comment by mookid8000** [comment]
It's because Rebus (at least from version 7, which is out as a prerelease now) is actually capable of using both.

You can switch to System.Text.Json by doing something like this:
```csharp
services.AddRebus(
    configure => configure
        .Transport(t => t.Use(...))
        .Serialization(s => s.UseSystemTextJson())
);
```
possibly passing `JsonSerializerOptions` as a parameter if you want to customize it.

~~And I am not sure Rebus will switch its default formatter away from Newtonsoft JSON, even though I would like to do so. The thing is, Newtonsoft has always been the default serialization in Rebus, and it's incredibly flexible~~

OK thinking some more about it, I could actually see Rebus switch to System.Text.Json from v7 as the default – I just need to verify that old Rebus instances running with Newtonsoft JSON in its default configuration will be fully interoperable with new Rebus instances running with System.Text.Json in its default configuration.

Do you know about any differences between the two? I know of a couple myself, but I am of course afraid to miss something.

**Comment by rdelgado-incomm** [comment]
Haha @mookid8000 I started digging and noticed that you guys had two so I suspected maybe it was legacy code or something.

So there are some differences, and they are noted here I believe: https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-6-0

There are some things that System.Text.Json doesn't do that Json.Net did automagically.  It handled enums better than System.Text.Json - or rather, it made assumptions and they worked for most devs.

There's a really nice table here outlining all of the differences: https://docs.microsoft.com/en-us/dotnet/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to?pivots=dotnet-6-0

The main benefit for switching to only System.Text.Json is that it is much more performant than Json.Net.  Our team at Incomm noticed some weird performance hits going on and Json.Net IIRC was creating a temporary file in memory or something that was a pretty significant performance hit for us.  We switched out our netcore services away from the default Json.Net text formatter and experienced pretty significant gains in performance in switching over.

**Comment by mookid8000** [comment]
That's some really great links, you're posting there, but it makes me realize that there's one huge difference between the two serializer, than cannot be ignored: How they handle .NET types and polymorphy!

Rebus defaults to configure Newtonsoft to include type information, so it embeds a `$type` field in each object, enabling deserialization of types with properties with abstract types and/or the use of interfaces!

But after having thought some more about it, I believe MOST users would not detect any change in behavior. And sometimes you have to introduce breaking changes, because the world moves forward, and so should we.

So I'll just go and conduct some experiments around which level of interoperability to expect, and then I'll see if we should simply make the switch in Rebus 7.

I'll be back!

**Comment by mookid8000** [comment]
Rebus 7.0.0-b03 is out now, and it's the first Rebus version to use System.Text.Json as the default serializer 😁 

I'm looking forward to seeing where this goes!

**Comment by mookid8000** [comment]
OK they're almost not interoperable at all 😐 when serializing collections with Newtonsoft JSON.NET and TypeNameHandling=All (as Rebus defaults to, which I really regret!), the JSON generated out of collections is actually an object with a field for the `$type` ,and then a field, `$values`, which holds the actual values.

Check this out: Class defined as

```csharp
record ComplexObject(string Name, int Age, decimal Height, IReadOnlyDictionary<string, string> Properties);
record ComplexMessage(string Label, ComplexObject[] Objects);
```

then this object

```csharp
new ComplexMessage("This is actually a bag of humanoids", new ComplexObject[]
{
    new(Name: "Møwns", Age: 42, Height: 181.5m, Properties: new Dictionary<string, string> {["Pockets"] = "full of stouts"}),
    new(Name: "Mørie", Age: 40, Height: 153.5m, Properties: new Dictionary<string, string> {["Beautiful"] = "indeed", ["Smart"] = "also"}),
})

```

serialized with both serializers results in the two following very very different pieces of JSON text:


```
JSON generated by SystemTextJsonSerializer: 

{
  "Label": "This is actually a bag of humanoids",
  "Objects": [
    {
      "Name": "Møwns",
      "Age": 42,
      "Height": 181.5,
      "Properties": {
        "Pockets": "full of stouts"
      }
    },
    {
      "Name": "Mørie",
      "Age": 40,
      "Height": 153.5,
      "Properties": {
        "Beautiful": "indeed",
        "Smart": "also"
      }
    }
  ]
}

JSON generated by JsonSerializer: 

{
  "$type": "Rebus.Tests.Serialization.TestJsonSerializerInteroperability+ComplexMessage, Rebus.Tests",
  "Label": "This is actually a bag of humanoids",
  "Objects": {
    "$type": "Rebus.Tests.Serialization.TestJsonSerializerInteroperability+ComplexObject[], Rebus.Tests",
    "$values": [
      {
        "$type": "Rebus.Tests.Serialization.TestJsonSerializerInteroperability+ComplexObject, Rebus.Tests",
        "Name": "Møwns",
        "Age": 42,
        "Height": 181.5,
        "Properties": {
          "$type": "System.Collections.Generic.Dictionary`2[[System.String, System.Private.CoreLib],[System.String, System.Private.CoreLib]], System.Private.CoreLib",
          "Pockets": "full of stouts"
        }
      },
      {
        "$type": "Rebus.Tests.Serialization.TestJsonSerializerInteroperability+ComplexObject, Rebus.Tests",
        "Name": "Mørie",
        "Age": 40,
        "Height": 153.5,
        "Properties": {
          "$type": "System.Collections.Generic.Dictionary`2[[System.String, System.Private.CoreLib],[System.String, System.Private.CoreLib]], System.Private.CoreLib",
          "Beautiful": "indeed",
          "Smart": "also"
        }
      }
    ]
  }
}

``` 

When trying to deserialize the Newtonsoft-generated JSON, it results in the following error:

```
  ----> System.Text.Json.JsonException : The JSON value could not be converted to Rebus.Tests.Serialization.TestJsonSerializerInteroperability+ComplexObject[]. Path: $.Objects | LineNumber: 0 | BytePositionInLine: 157.
```

which I'll just leave here for Bingleability.

So this should be noted: Rebus v7 has switched to use System.Text.Json as its default JSON serializer, which results in much more compact JSON, but also less options for message modeling.

Interoperability with Rebus v6 and all versions prior to that can be had by switching back to Newtonsoft JSON.NET, which is still included together with the Rebus NuGet package, and can be done simply by doing something like this:

```csharp
services.AddRebus(
	configure => configure
		.(...)
		.Serializer(s => s.UseNewtonsoftJson()) // 👈 used to be the default
);
```

**Comment by mookid8000** [comment]
I've released Rebus 7.0.0-b04 now with some minor changes (allow trailing commas and comments in JSON)
